File: thirdpart/ethereum/src/main/java/rehanced/com/simpleetherwallet/network/EtherscanAPI.java
package rehanced.com.simpleetherwallet.network;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.BitmapDrawable;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.List;
import java.util.concurrent.TimeUnit;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Protocol;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;
import rehanced.com.simpleetherwallet.interfaces.LastIconLoaded;
import rehanced.com.simpleetherwallet.interfaces.StorableWallet;
import rehanced.com.simpleetherwallet.utils.RequestCache;
import rehanced.com.simpleetherwallet.utils.TokenIconCache;

@Deprecated
public class EtherscanAPI {

    private static final String OFFICIAL_HOST = "http://api.etherscan.io/";
    private static final String RINKEBY_HOST = "http://rinkeby.etherscan.io/";

    private String token;
    private String currentHost;//当前域名

    private static EtherscanAPI instance;

    public static EtherscanAPI getInstance() {
        if (instance == null) {
            instance = new EtherscanAPI();
        }
        return instance;
    }

    /**
     * transfer network id related to
     *
     * @return
     */
    public static Byte getChainId() {
        //        if(BuildConfig.DEBUG) {
        //            return (byte)3;
        //        }else {
        //            return (byte)1;
        //        }
        return (byte) 1;
    }

    public void getPriceChart(long starttime, int period, boolean usd, Callback b) throws IOException {
        get(true, "http://poloniex.com/public?command=returnChartData&currencyPair=" + (usd ? "USDT_ETH" : "BTC_ETH") + "&start=" + starttime + "&end=9999999999&period=" + period, b);
    }


    /**
     * Retrieve all internal transactions from address like contract calls, for normal transactions @see rehanced.com.simpleetherwallet.network.EtherscanAPI#getNormalTransactions() )
     *
     * @param address Ether address
     * @param b       Network callback to @see rehanced.com.simpleetherwallet.fragments.FragmentTransactions#update() or @see rehanced.com.simpleetherwallet.fragments.FragmentTransactionsAll#update()
     * @param force   Whether to force (true) a network call or use cache (false). Only true if user uses swiperefreshlayout
     * @throws IOException Network exceptions
     */
    public void getInternalTransactions(String address, Callback b, boolean force) throws IOException {
        if (!force && RequestCache.getInstance().contains(RequestCache.TYPE_TXS_INTERNAL, address)) {
            b.onResponse(null, new Response.Builder().code(200).message("").request(new Request.Builder()
                    .url("api?module=account&action=txlistinternal&address=" + address + "&startblock=0&endblock=99999999&sort=asc&apikey=" + token)
                    .build()).protocol(Protocol.HTTP_1_0).body(ResponseBody.create(MediaType.parse("JSON"), RequestCache.getInstance().get(RequestCache.TYPE_TXS_INTERNAL, address))).build());
            return;
        }
        get("api?module=account&action=txlistinternal&address=" + address + "&startblock=0&endblock=99999999&sort=asc&apikey=" + token, b);
    }


    /**
     * Retrieve all normal ether transactions from address (excluding contract calls etc, @see rehanced.com.simpleetherwallet.network.EtherscanAPI#getInternalTransactions() )
     *
     * @param address Ether address
     * @param b       Network callback to @see rehanced.com.simpleetherwallet.fragments.FragmentTransactions#update() or @see rehanced.com.simpleetherwallet.fragments.FragmentTransactionsAll#update()
     * @param force   Whether to force (true) a network call or use cache (false). Only true if user uses swiperefreshlayout
     * @throws IOException Network exceptions
     */
    public void getNormalTransactions(String address, Callback b, boolean force) throws IOException {
        if (!force && RequestCache.getInstance().contains(RequestCache.TYPE_TXS_NORMAL, address)) {
            b.onResponse(null, new Response.Builder().code(200).message("").request(new Request.Builder()
                    .url("api?module=account&action=txlist&address=" + address + "&startblock=0&endblock=99999999&sort=asc&apikey=" + token)
                    .build()).protocol(Protocol.HTTP_1_0).body(ResponseBody.create(MediaType.parse("JSON"), RequestCache.getInstance().get(RequestCache.TYPE_TXS_NORMAL, address))).build());
            return;
        }
        get("api?module=account&action=txlist&address=" + address + "&startblock=0&endblock=99999999&sort=asc&apikey=" + token, b);
    }


    public void getEtherPrice(Callback b) throws IOException {
        get("api?module=stats&action=ethprice&apikey=" + token, b);
    }


    public void getGasPrice(Callback b) throws IOException {
        get("api?module=proxy&action=eth_gasPrice&apikey=" + token, b);
    }


    /**
     * Get token balances via ethplorer.io
     *
     * @param address Ether address
     * @param b       Network callback to @see rehanced.com.simpleetherwallet.fragments.FragmentDetailOverview#update()
     * @param force   Whether to force (true) a network call or use cache (false). Only true if user uses swiperefreshlayout
     * @throws IOException Network exceptions
     */
    public void getTokenBalances(String address, Callback b, boolean force) throws IOException {
        if (!force && RequestCache.getInstance().contains(RequestCache.TYPE_TOKEN, address)) {
            b.onResponse(null, new Response.Builder().code(200).message("").request(new Request.Builder()
                    .url("https://api.ethplorer.io/getAddressInfo/" + address + "?apiKey=freekey")
                    .build()).protocol(Protocol.HTTP_1_0).body(ResponseBody.create(MediaType.parse("JSON"), RequestCache.getInstance().get(RequestCache.TYPE_TOKEN, address))).build());
            return;
        }
        get("http://api.ethplorer.io/getAddressInfo/" + address + "?apiKey=freekey", b);
    }


    /**
     * Download and save token icon in permanent image cache (TokenIconCache)
     *
     * @param c         Application context, used to load TokenIconCache if reinstanced
     * @param tokenName Name of token
     * @param lastToken Boolean defining whether this is the last icon to download or not. If so callback is called to refresh recyclerview (notifyDataSetChanged)
     * @param callback  Callback to @see rehanced.com.simpleetherwallet.fragments.FragmentDetailOverview#onLastIconDownloaded()
     * @throws IOException Network exceptions
     */
    public void loadTokenIcon(final Context c, String tokenName, final boolean lastToken, final LastIconLoaded callback) throws IOException {
        if (tokenName.indexOf(" ") > 0) {
            tokenName = tokenName.substring(0, tokenName.indexOf(" "));
        }
        if (TokenIconCache.getInstance(c).contains(tokenName)) {
            return;
        }

        if (tokenName.equalsIgnoreCase("OMGToken")) {
            tokenName = "omise";
        } else if (tokenName.equalsIgnoreCase("0x")) {
            tokenName = "0xtoken_28";
        }

        final String tokenNamef = tokenName;
        get(true, "http://etherscan.io//token/images/" + tokenNamef + ".PNG", new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                if (c == null) {
                    return;
                }
                ResponseBody in = response.body();
                InputStream inputStream = in.byteStream();
                BufferedInputStream bufferedInputStream = new BufferedInputStream(inputStream);
                final Bitmap bitmap = BitmapFactory.decodeStream(bufferedInputStream);
                TokenIconCache.getInstance(c).put(c, tokenNamef, new BitmapDrawable(c.getResources(), bitmap).getBitmap());
                // if(lastToken) // TODO: resolve race condition
                callback.onLastIconDownloaded();
            }
        });
    }


    public void getGasLimitEstimate(String to, Callback b) throws IOException {
        get("api?module=proxy&action=eth_estimateGas&to=" + to + "&value=0xff22&gasPrice=0x051da038cc&gas=0xffffff&apikey=" + token, b);
    }


    public void getBalance(String address, Callback b) throws IOException {
        get("api?module=account&action=balance&address=" + address + "&apikey=" + token, b);
    }


    public void getNonceForAddress(String address, Callback b) throws IOException {
        get("api?module=proxy&action=eth_getTransactionCount&address=" + address + "&tag=latest&apikey=" + token, b);
    }


    public void getPriceConversionRates(String currencyConversion, Callback b) throws IOException {
        get(true, "https://api.fixer.io/latest?base=USD&symbols=" + currencyConversion, b);
    }


    public void getBalances(List<StorableWallet> addresses, Callback b) throws IOException {
        StringBuilder urlBuilder = new StringBuilder("api?module=account&action=balancemulti&address=");
        for (StorableWallet address : addresses) {
            urlBuilder.append(address.getPubKey());
            urlBuilder.append(",");
        }
        if (!addresses.isEmpty()) {
            urlBuilder.deleteCharAt(urlBuilder.length() - 1);
        }
        // remove last , AND add token
        urlBuilder.append("&tag=latest&apikey=");
        urlBuilder.append(token);
        get(urlBuilder.toString(), b);
    }


    public void forwardTransaction(String raw, Callback b) throws IOException {
        get("api?module=proxy&action=eth_sendRawTransaction&hex=" + raw + "&apikey=" + token, b);
    }


    public void get(String url, Callback b) throws IOException {
        get(false, url, b);
    }

    public void get(boolean useFullUrl, String url, Callback b) throws IOException {
        Request request = new Request.Builder()
                .url(useFullUrl ? url : currentHost + url)
                .build();
        OkHttpClient client = new OkHttpClient.Builder()
                .connectTimeout(10, TimeUnit.SECONDS)
                .writeTimeout(10, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .build();

        client.newCall(request).enqueue(b);
    }

    private EtherscanAPI() {
        token = "UGDGCQ1J92XXTUYKFEMEZWJZZFCNTP9A54";
        //        currentHost = BuildConfig.DEBUG ? RINKEBY_HOST : OFFICIAL_HOST;
        currentHost = OFFICIAL_HOST;
    }

}

File: wallet/src/main/java/com/bcm/messenger/wallet/network/EtherscanAPI.kt
package com.bcm.messenger.wallet.network

import com.bcm.messenger.common.utils.AppUtil
import com.orhanobut.logger.Logger
import com.bcm.messenger.utility.logger.ALog
import com.bcm.messenger.wallet.utils.BCMWalletManager
import java.io.IOException
import okhttp3.Callback
import okhttp3.MediaType
import okhttp3.Protocol
import okhttp3.Request
import okhttp3.Response
import okhttp3.ResponseBody

/**
 * ETH 平台接口api类
 * Created by wjh on 2018/05/19
 */
class EtherscanAPI private constructor() {

    companion object {

        private val TAG = "EtherscanAPI"

        private val OFFICIAL_HOST = "https://api.etherscan.io/"
        private val RINKEBY_HOST = "https://api-rinkeby.etherscan.io/"
//                "http://rinkeby.etherscan.io/"

        val INSTANCE: EtherscanAPI by lazy { EtherscanAPI() }

        val chainId: Byte
            get() = if (AppUtil.useDevBlockChain()) {
                4.toByte()
            } else {
                1.toByte()
            }
    }

    private val token: String
    private val currentHost: String//当前域名

    init {
        token = "UGDGCQ1J92XXTUYKFEMEZWJZZFCNTP9A54"
        currentHost = if (AppUtil.useDevBlockChain()) RINKEBY_HOST else OFFICIAL_HOST
    }

    fun getPriceChart(startTime: Long, period: Int, usd: Boolean, b: Callback) {
        request(true, "http://poloniex.com/public?command=returnChartData&currencyPair=" + (if (usd) "USDT_ETH" else "BTC_ETH") + "&start=" + startTime + "&end=9999999999&period=" + period, b)
    }

    fun getPriceChart(startTime: Long, period: Int, usd: Boolean): Response {
        return request(true, "http://poloniex.com/public?command=returnChartData&currencyPair=" + (if (usd) "USDT_ETH" else "BTC_ETH") + "&start=" + startTime + "&end=9999999999&period=" + period)
    }

    /**
     * 查询账号余额
     */
    fun getBalance(address: String, b: Callback) {
        request("api?module=account&action=balance&address=$address&apikey=$token", b)
    }

    fun getBalance(address: String, useCache: Boolean = false): String {
        //首先判断是否强制刷新，如果不是可以从缓存读取
        if(useCache && RequestCache.instance.contains(RequestCache.TYPE_BALANCES, address)) {
            return RequestCache.instance.get(RequestCache.TYPE_BALANCES, address) ?: ""
        }
        val response = request(false, "api?module=account&action=balance&address=$address&apikey=$token")
        return if(response.isSuccessful) {
            val result = response.body()?.string() ?: ""
            //更新缓存内容
            RequestCache.instance.put(RequestCache.TYPE_BALANCES, address, result)
            result

        }else {
            throw Exception("getBalance fail: ${response.code()}")
        }
    }

    /**
     * 查询多个地址的余额
     */
    fun getBalances(addresses: List<String>, b: Callback) {
        val urlBuilder = StringBuilder("api?module=account&action=balancemulti&address=")
        for (address in addresses) {
            urlBuilder.append(address)
            urlBuilder.append(",")
        }
        if (!addresses.isEmpty()) {
            urlBuilder.deleteCharAt(urlBuilder.length - 1)
        }
        // remove last , AND add token
        urlBuilder.append("&tag=latest&apikey=")
        urlBuilder.append(token)
        request(urlBuilder.toString(), b)
    }

    fun getBalances(addresses: List<String>): String {
        val urlBuilder = StringBuilder("api?module=account&action=balancemulti&address=")
        for (address in addresses) {
            urlBuilder.append(address)
            urlBuilder.append(",")
        }
        if (!addresses.isEmpty()) {
            urlBuilder.deleteCharAt(urlBuilder.length - 1)
        }
        // remove last , AND add token
        urlBuilder.append("&tag=latest&apikey=")
        urlBuilder.append(token)
        val response= request(false, urlBuilder.toString())
        return if(response.isSuccessful) {
            response.body()?.string() ?: ""
        }else {
            throw Exception("getBalances fail: ${response.code()}")
        }
    }

    /**
     * Retrieve all internal transactions from address like contract calls, for normal transactions @see rehanced.com.simpleetherwallet.network.EtherscanAPI#getNormalTransactions() )
     *
     * @param address Ether address
     * @param b       Network callback to @see rehanced.com.simpleetherwallet.fragments.FragmentTransactions#update() or @see rehanced.com.simpleetherwallet.fragments.FragmentTransactionsAll#update()
     * @param force   Whether to force (true) a network call or use cache (false). Only true if user uses swiperefreshlayout
     * @throws IOException Network exceptions
     */
    fun getInternalTransactions(address: String, b: Callback, useCache: Boolean) {
        var cacheString: String? = null
        if (useCache && RequestCache.instance.contains(RequestCache.TYPE_TXS_INTERNAL, address)) {
            cacheString = RequestCache.instance.get(RequestCache.TYPE_TXS_INTERNAL, address)
        }
        request("api?module=account&action=txlistinternal&address=$address&startblock=0&endblock=99999999&sort=asc&apikey=$token", b, cacheString)
    }

    fun getTransactions(normal: Boolean, address: String, useCache: Boolean = false): String {
        val urlPart = if(normal) {
            "api?module=account&action=txlist&address=$address&startblock=0&endblock=99999999&sort=desc&apikey=$token"
        }else {
            "api?module=account&action=txlistinternal&address=$address&startblock=0&endblock=99999999&sort=desc&apikey=$token"
        }
        val type = if(normal) RequestCache.TYPE_TXS_NORMAL else RequestCache.TYPE_TXS_INTERNAL
        if(useCache && RequestCache.instance.contains(type, address)) {//首先判断是否强制刷新，如果不是可以直接从缓存读取内容
            return RequestCache.instance.get(type, address) ?: ""
        }
        val response = request(false, urlPart)
        return if(response.isSuccessful) {
            val result = response.body()?.string() ?: ""
            //更新缓存内容
            RequestCache.instance.put(type, address, result)
            result
        }else {
            throw Exception("getInternalTransactions fail: ${response.code()}")
        }
    }

    /**
     * Retrieve all normal ether transactions from address (excluding contract calls etc, @see rehanced.com.simpleetherwallet.network.EtherscanAPI#getInternalTransactions() )
     *
     * @param address Ether address
     * @param b       Network callback to @see rehanced.com.simpleetherwallet.fragments.FragmentTransactions#update() or @see rehanced.com.simpleetherwallet.fragments.FragmentTransactionsAll#update()
     * @param force   Whether to force (true) a network call or use cache (false). Only true if user uses swiperefreshlayout
     * @throws IOException Network exceptions
     */
    fun getNormalTransactions(address: String, b: Callback, useCache: Boolean) {
        var cacheString: String? = null
        if (useCache && RequestCache.instance.contains(RequestCache.TYPE_TXS_NORMAL, address)) {
            cacheString = RequestCache.instance.get(RequestCache.TYPE_TXS_NORMAL, address)
        }
        request("api?module=account&action=txlist&address=$address&startblock=0&endblock=99999999&sort=asc&apikey=$token", b, cacheString)
    }

    /**
     * 发送交易
     */
    fun forwardTransaction(raw: String, b: Callback) {
        request("api?module=proxy&action=eth_sendRawTransaction&hex=$raw&apikey=$token", b)
    }

    fun forwardTransaction(raw: String): Response {
        return request(false, "api?module=proxy&action=eth_sendRawTransaction&hex=$raw&apikey=$token")
    }

    fun getEtherPrice(b: Callback) {
        request("api?module=stats&action=ethprice&apikey=$token", b)
    }

    fun getEtherPrice(): Response {
        return request(false, "api?module=stats&action=ethprice&apikey=$token")
    }

    fun getGasPrice(b: Callback) {
        request("api?module=proxy&action=eth_gasPrice&apikey=$token", b)
    }

    fun getGasPrice(): Response {
        return request(false, "api?module=proxy&action=eth_gasPrice&apikey=$token")
    }

    /**
     * Get token balances via ethplorer.io
     *
     * @param address Ether address
     * @param b       Network callback to @see rehanced.com.simpleetherwallet.fragments.FragmentDetailOverview#update()
     * @param force   Whether to force (true) a network call or use cache (false). Only true if user uses swiperefreshlayout
     * @throws IOException Network exceptions
     */
    fun getTokenBalances(address: String, b: Callback, force: Boolean) {
        var cacheString: String? = null
        if (!force && RequestCache.instance.contains(RequestCache.TYPE_TOKEN, address)) {
            cacheString = RequestCache.instance.get(RequestCache.TYPE_TOKEN, address)
        }
        request("http://api.ethplorer.io/getAddressInfo/$address?apiKey=freekey", b, cacheString)

    }

    fun getTokenBalances(address: String): Response {
        return request(true, "http://api.ethplorer.io/getAddressInfo/$address?apiKey=freekey")
    }

    /**
     * 查询小费gasLimit
     */
    fun getGasLimitEstimate(to: String, b: Callback) {
        request("api?module=proxy&action=eth_estimateGas&to=$to&apikey=$token", b)
    }

    fun getGasLimitEstimate(to: String): Response {
        return request(false, "api?module=proxy&action=eth_estimateGas&to=$to&apikey=$token")
    }

    /**
     * 查询最合适的gasPrice
     */
    fun getSuggestGasPrice(b: Callback) {
        request(true, "http://ethgasstation.info/json/ethgasAPI.json", b)
    }

    fun getSuggestGasPrice(): String {
        val response = request(true, "http://ethgasstation.info/json/ethgasAPI.json")
        return if(response.isSuccessful) {
            response.body()?.string() ?: ""
        }else {
            throw Exception("getSuggestGasPrice fail: ${response.code()}")
        }
    }

    /**
     * 查询nonce
     */
    fun getNonceForAddress(address: String, b: Callback) {
        request("api?module=proxy&action=eth_getTransactionCount&address=$address&tag=latest&apikey=$token", b)
    }

    fun getNonceForAddress(address: String): Response {
        return request(false, "api?module=proxy&action=eth_getTransactionCount&address=$address&tag=latest&apikey=$token")
    }

    /**
     * 查询汇率
     */
    fun getPriceConversionRates(currencyConversion: String, b: Callback) {
        request(true, "https://api.fixer.io/latest?base=USD&symbols=$currencyConversion", b)
    }

    fun getPriceConversationRates(currencyConversion: String): Response {
        return request(true, "https://api.fixer.io/latest?base=USD&symbols=$currencyConversion")
    }


    private fun request(url: String, b: Callback, cacheString: String? = null) {
        request(false, url, b, cacheString)
    }

    private fun request(useFullUrl: Boolean, url: String, b: Callback, cacheString: String? = null) {
        val fullUrl = if (useFullUrl) url else currentHost + url
        val request = Request.Builder()
                .url(fullUrl)
                .build()
        val call = BCMWalletManager.provideHttpClient().newCall(request)
        if (!cacheString.isNullOrEmpty()) {
            Logger.d("EtherscanAPI直接从缓存获取数据")
            b.onResponse(call, Response.Builder().code(200).message("").request(Request.Builder()
                    .url(fullUrl).build()).protocol(Protocol.HTTP_1_0).body(ResponseBody.create(MediaType.parse("JSON"), cacheString)).build())
            return
        } else {
            call.enqueue(b)
        }
    }

    private  fun request(useFullUrl: Boolean, url: String): Response {
        val fullUrl = if (useFullUrl) url else currentHost + url
        val request = Request.Builder()
                .url(fullUrl)
                .build()
        ALog.d(TAG, "request url: $fullUrl")
        val client = BCMWalletManager.provideHttpClient()
        return client.newCall(request).execute()
    }



}
